### PR TITLE
Project 'Support for sequence profiles'

### PR DESCRIPTION
**Pull request by MaxGreil**
This PR aims to implement project 'Support for sequence profiles'

- Implementation of new new class SequenceProfile
- Implementation of tests

@padix-key can you take a look?

### CODE REVIEW

File: src/biotite/sequence/profile.py
Comment (padix-key) [inline src/biotite/sequence/profile.py:55]: I suggest that the public access is handled via `@property`. For example I think the user should not be able to change the alphabet, because then the size of the frequency matrix does not fit anymore. Instead I think `position_frequency_matrix` and `position_frequency_matrix` should check whether the shape is still the same, when these attributes are set.

Furthermore, the variable names `position_frequency_matrix` and `position_frequency_matrix` are quite long in my opinion. May I suggest `symbols` and `gaps` instead?

File: src/biotite/sequence/profile.py
Comment (padix-key) [inline src/biotite/sequence/profile.py:76]: Could you elaborate in the docstring how an alignment ins converted into a profile?

File: src/biotite/sequence/profile.py
Comment (padix-key) [inline src/biotite/sequence/profile.py:78]: Using `get_gapped_sequences()` in this place is quite inefficient: The trace is converted to symbol codes, the symbol codes to symbols and the symbols into a string, and vice versa.

Probably `sequence.align.get_codes()` (https://www.biotite-python.org/apidoc/biotite.sequence.align.get_codes.html) is more efficient in this place. Then you can use `np.bincount()`for every column. You could handle `-1` values by replacing them with `len(alphabet)`

File: src/biotite/sequence/profile.py
Comment (padix-key) [inline src/biotite/sequence/profile.py:79]: Could you check if all sequences in the alignment use the same alphabet?

File: src/biotite/sequence/profile.py
Comment (padix-key) [inline src/biotite/sequence/profile.py:111]: The current implementation runs this for every position in the profile. However `np.where()` and `np.amax()` are also able to handle multidimensional arrays. So maybe it is possible to run this once for the entire profile to increase the efficiency.

File: src/biotite/sequence/profile.py
Comment (padix-key) [inline src/biotite/sequence/profile.py:98]: I think it is more consistent with the rest of the package if this function returns the consensus as `Sequence` object, either as `NucleotideSequence`, `ProteinSequence` or `GeneralSequence` (for custom alphabets) object. Maybe a parameter like `as_general` would be helpful, if the user wants a `GeneralSequence` in any case.

File: tests/sequence/test_profile.py
Comment (padix-key) [inline tests/sequence/test_profile.py:37]: For completeness this function should also test the case where the resulting sequence contains an ambiguous symbol.

**Review by padix-key — CHANGES_REQUESTED** [review]
Hi, thank you for the PR. Here is my review. I especially like the idea that the consensus sequence contains unambiguous symbols, if the input sequences have two different symbols at the same frequency.

File: src/biotite/sequence/profile.py
Comment (MaxGreil) [reply src/biotite/sequence/profile.py:157]: Hi @padix-key , I chose this implementation because I need to get multiple maximum values for each row of ndarray symbols. I did not find a solution to to this once for a multidimensional array. Do you have an idea how to do this?

**Comment by MaxGreil** [comment]
Hi @padix-key , I applied your requested changes. Can you have another look?

File: src/biotite/sequence/profile.py
Comment (padix-key) [reply src/biotite/sequence/profile.py:157]: For `np.where()` you are probably right, I also have currently no direct solution to solve this in a single function call. However, you could at least obtain the maximum value in each column with `np.max(self._symbols, axis=1)`

File: src/biotite/sequence/profile.py
Comment (padix-key) [inline src/biotite/sequence/profile.py:158]: Calling `self.alphabet.get_symbols()` for each column introduces some overhead, since the list of symbols is copied everytime. it is probably more efficient to use `Alphabet.decode()` instead or, even better, to directly use symbols codes and skip using characters entirely.

**Review by padix-key — CHANGES_REQUESTED** [review]
Hi, thank you for the changes. I left some further comments. Although I think most of the code will already work as intended, there are still some issues regarding the performance in my opinion.

**Comment by MaxGreil** [comment]
Hi @padix-key , I applied the changes. Can you have another look?

File: src/biotite/sequence/profile.py
Comment (padix-key) [inline src/biotite/sequence/profile.py:202]: See above

File: src/biotite/sequence/profile.py
Comment (padix-key) [inline src/biotite/sequence/profile.py:215]: See above

File: src/biotite/sequence/profile.py
Comment (padix-key) [inline src/biotite/sequence/profile.py:44]: Can you use a tuple of integers instead of a string as dictionary key? This would remove the time-costly integer-to-string conversion. Furthermore, I think that a tuple of integers makes it more clear that they key is based on sequence codes.

File: src/biotite/sequence/profile.py
Comment (padix-key) [inline src/biotite/sequence/profile.py:162]: I think, this is basically a *no-op* you can omit.

**Review by padix-key — COMMENTED** [review]
Thanks for the additional work. As a final change, i would suggest using tuples as dictionary keys for getting the consensus sequence. After this I would merge this PR, if you have no objections.

**Comment by MaxGreil** [comment]
Hi @padix-key , I changed to tuples. I have no objects to merge this PR. I will then add the missing features for project sequence profiles in another PR.

**Review by padix-key — APPROVED** [review]
OK, thank you for your contribution! :+1:

**Pull request merged**
